thefuck/rules/wrong_hyphen_before_subcommand.py

```diff
@@ -0,0 +1,151 @@
+    :type number: int (optional)
+    :rtype: [str]
+    """
+    path = get_manpage_path(command, number)
+    if path:
+        with gzip.open(path, "r") as f:
+            return list(filter(None, (textify(line) for line in f.readlines())))
+    else:
+        return []
+
+
+def textify(manline):
+    """Stringifies a manpage line (strips markdown instructions, unescapes characters and removes quotes)
+
+    :type manline: bytes
+    :rtype: str
+    """
+    # Stripping line headers .SH, .B
+    manline = re.sub(r"^\.[A-Za-z]+( |$)", "", str(manline, "utf-8").strip())
+
+    # Stripping format
+    manline = re.sub(r"\\f.", "", manline)
+    manline = re.sub(r"\\(.)", r"\1", manline)
+
+    # Stripping quotes
+    return manline.replace('"', '').replace('\'', '')
+
+
+def get_synopsis(man):
+    """Given a parsed manpage entry (as returned from get_manpage), returns the synopsis line from the manpage
+
+    :type man: [str]
+    :rtype: str
+    """
+    try:
+        return " ".join(line for line in man[man.index("SYNOPSIS") + 1: man.index("DESCRIPTION")])
+    except ValueError:
+        return ""
+
+
+def get_flags(man):
+    """Given a parsed manpage entry (as returned from get_manpage), returns a list of lines which contain flag information
+
+    :type man: [str]
+    :rtype: [str]
+    """
+    return [x for x in man if x[0] == '-']
+
+
+def find_at_start(word, line):
+    return line.startswith(word) and\
+        (len(line) == len(word) or line[len(word)] in ', (:')
+
+
+def find_subcommand(manpage, command, subcommand):
+    return any(map(partial(find_at_start, subcommand), manpage)) or\
+        any(map(partial(find_at_start, "%s-%s" % (command, subcommand)), manpage))
+
+
+def find_flag(flags, flag):
+    return any(map(partial(find_at_start, '-' + flag), flags))
+
+
+@sudo_support
+def match(command):
+    if command.script_parts[0] in get_all_executables():
+        return False
+
+    if '-' not in command.script_parts[0]:
+        return False
+
+    cmd, subcmd = command.script_parts[0].split('-')[:2]
+    if cmd not in get_all_executables():
+        return False
+
+    manpage = get_manpage(cmd)
+    if not manpage:
+        return False
+
+    synopsis = get_synopsis(manpage)
+    if subcmd in synopsis:
+        return True
+
+    if find_subcommand(manpage, cmd, subcmd):
+        return True
+
+    flags = get_flags(manpage)
+    for flag in subcmd:
+        if not find_flag(flags, flag):
+            return False
+
+    return True
+
+
+@sudo_support
+def get_new_command(command):
+    cmd, subcmd = command.script_parts[0].split('-')[:2]
+    manpage = get_manpage(cmd)
+    synopsis = get_synopsis(manpage)
+
+    if subcmd in synopsis or find_subcommand(manpage, cmd, subcmd):
+        # We are dealing with an accidentally added hyphen
+        return command.script.replace('-', ' ', 1)
+
+    else:
+        # We are dealing with a missing space
+        return command.script.replace('-', ' -', 1)
```

Comment:
Perhaps a simpler version of the entire rule would be better.

IMO, TheFuck should delegate that choice to the user.

Following the path the current rule is going, the failed command `apt-install foo` would get two suggestions, `apt -install foo` and `apt install foo`, among others.

Thing is that the former – `apt -install foo` – already gets suggested by the existing `missing_space_before_subcommand ` rule.

Looking at it that way, all the `wrong_hyphen_before_subcommand` – or even an *augmented* `missing_space_before_subcommand` – rule would have to do is replace the hyphen with a space.

What you think?

Reply:
Following your suggestions, I tried to simplify the rule down. I'm afraid I over-simplified it (since it may now suggest `git hello` as a replacement to `git-hello` even though `hello` is not a legitimate subcommand for `git`).

I'd be happy to hear your opinion about that. Should rules do above and beyond to verify their command suggestions? Or should the current way of command checking be the norm?

Anyway, another issue I have is that the rule now should not try to handle cases such as `ls-la` (-> `ls -la`), but as it doesn't check the manpages anymore there is no way for the rule to determine if `la` is a subcommand or flags. Thus it suggests `ls la`. Now the issue is that the priority of the rule (as I see it) should be lower than `no_command`'s priority, and the `missing_space_before_subcommand` is higher than `no_command`'s. Thus the `wrong_hyphen_before_subcommand` shadows the `missing_space_before_subcommand` rule.

Any suggestions?

Reply:
I think it's better if the rule does not try to be super smart. In the same sense as `missing_space_before_subcommand`. Let it suggest any fix that has some probability to work. The user decides to accept. That's my own opinion.

Regarding the `priority`, I think it should be set to `3900`. It's more close to `missing_space_before_subcommand` than to `no_command`.

@nvbn wanna chime in, pal?